---
title: "TimeCourse analysis of Treated/Not treated Controls and Patients \nin D0,D1,D3 for Femoral Neck fractures"
author: "Maria Kondili"
date: "24/05/2021,edited: 16/07/21"
output: html_document
editor_options:
  
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

```{r,load_libraries}  
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(tidyr))
# BiocManager::install("maSigPro")
library(maSigPro)

```


```{r}
## Doc@:
##  https://bioconductor.org/packages/release/bioc/vignettes/maSigPro/inst/doc/maSigProUsersGuide.pdf

##Other package :
# Futschik M: Mfuzz: Soft clustering of time series gene expression data 2007 
# [http://itb.biologie.hu-berlin.de/~futschik/software/R/Mfuzz/index.html].

```


```{r set_wd}

proj_dir <- "/shared/projects/secretome/Col_Femoral/"
data_dir <- paste0(proj_dir,"Data/")
workdir  <- paste0(proj_dir,"maSigPro_v2/")

setwd(workdir)

```


```{r read_data,echo=F,include=F}

affy_data_74 <- read.delim(paste0(data_dir, 
                        "ExpressionDataNormaliz_Affymetrix_ClariomsHuman_74samples_18627id_edMK.txt"), 
                        header=TRUE, as.is=TRUE)

# rownames(affy_data_74) <- affy_data_74 $EntrezGeneID

## Remove "HIP" prefix on some samples ,to match with Patients listing table 
pos_hip_affy <- grep("HIP",colnames(affy_data_74))
colnames(affy_data_74)[pos_hip_affy] <- gsub("HIP","",colnames(affy_data_74)[pos_hip_affy] )


## patients assignment to groups :: profile description : 
patients_list <- read.delim(paste0(data_dir,"Listing_patients_transcriptome_edMK.tsv"),
                            header=TRUE, as.is=TRUE)

#patients_ids <- paste0("HIP", gsub("^[0-7]*._", "" ,patients_list$FileID), sep="") ##> Like l2fc_transf_data 

patients_list$FileID <- paste0("X",patients_list$FileID)
patients_list %>% head


```


```{r correct_patient_id,echo=F,include=F }

pos_cd14 <- grep("CD14_",patients_list$FileID )
patients_list$FileID[pos_cd14] <- gsub("CD14_", "", patients_list$FileID[pos_cd14])

pos_hip <- grep("HIP",patients_list$FileID)
patients_list$FileID[pos_hip]  <- gsub("HIP", "", patients_list$FileID[pos_hip])


```


```{r find_common_ids,echo=F,include=F }

# patients_list$FileID[which(! patients_list$FileID %in% colnames(affy_data_74))]
patients_list <- patients_list[-which(! patients_list$FileID %in% colnames(affy_data_74)), ]

#  which(! colnames(affy_data_74)  %in% patients_list$FileID)[-1]  #>>[1] 72 73
# colnames(affy_data_74)[72:73]
# [1] "X718_J0_NS"  "X728_J0_LPS"
affy_data_74  <- affy_data_74[, -which(! colnames(affy_data_74) %in% patients_list$FileID)[-1]]

###> save new tables : 
write_tsv(patients_list, file=paste0(data_dir,"Listing_patients_transcriptome_modifMK.tsv"),col_names = T)

write_tsv(affy_data_74,
          file=paste0(data_dir,"ExpressionDataNormaliz_Affymetrix_ClariomsHuman_74samples_18627id_modifMK.tsv"),
          col_names = T)


```

Using Martin's transf.data :  re-analyse taking as INPUT LOG2(LPS/NS) ,and adding CTRL(non_fractured)
Here are clinical patient_info + counts per gene in columns

Transformation from raw shown in script: transform_raw_data_by_MartinLarsen.R

```{r extract_expr, echo=T, include=T }

##> log2(LPS/NS) of affy_data : 
log2fc_data <- read.table(paste0(data_dir,"log2FC_transformed_data_total_MartinLarsen.tsv"), 
                        sep="\t",header=T)

samplesheet <- log2fc_data[,1:7] # clinical samplesheet only
##> 37 patients

#> Transpose to have genes in rows 
log2fc_expr <- t(log2fc_data[,8:ncol(log2fc_data)])

## to obtain positive values (shift gaussian distrib to +10 ) : 
log2fc_expr <- log2fc_expr + 10


##> Arrange samples nomenclature: 

samplesheet$Group %>% as.factor() %>% levels

row.names(samplesheet) <- paste( paste0("X",samplesheet$ID),
                                          gsub("HIP","",samplesheet$PatientCode),
                                          samplesheet$TimePointClin, sep="_")

names_wo_0 <- grep("^X[1-9]_", row.names(samplesheet))
row.names(samplesheet)[names_wo_0] <- gsub("X", "X0", row.names(samplesheet)[names_wo_0])

ctrl_samples <- rownames(samplesheet)[samplesheet$Group=="CTR"]

rownames(samplesheet)[which(samplesheet$Group=="CTR")] <- gsub("*_J[0-9]$|_CTR","_CTRL", ctrl_samples)

colnames(log2fc_expr) <- rownames(samplesheet)

###>save in file : 
#write.table(samplesheet_l2fcData, paste0(data_dir,"samplesheet_log2fc_lps_ns.tsv"),col.names = T,row.names = T)



```


#### Create edesign for masigPro, by the patients-id and samplesheet info : 

```{r edesign_with_LPS/NS_samples }

## Created following file manually, by looking the samples of "samplesheet" for J0-J1-J3 & CTRL.
# Ignored J2,J4,J6,J8--> rescaled J2,J4 in J1,J3
## Replicates column is designed for every timepoint ,to change according to Mort / Vivant group
## Ignore NS, LPS, because Counts contain ratio of NS/LPS -> will compare only Mort-vs-Viv to draw conclusions

edesign <- read_delim(paste0(workdir,"edesign_Timepoint_sorted_31_samples.tsv"),
                      delim="\t", col_names = T, show_col_types = FALSE)

```

RENAME samples of J2,J6,J7,J8 from log2fc_data :

```{r}

bad_timing <- which(! colnames(log2fc_expr) %in% edesign$Description )

colnames(log2fc_expr)[bad_timing]<- gsub("_J1","_J0", colnames(log2fc_expr)[bad_timing])
colnames(log2fc_expr)[bad_timing]<- gsub("_J2","_J1",  colnames(log2fc_expr)[bad_timing])
colnames(log2fc_expr)[bad_timing]<- gsub("_J4","_J3",  colnames(log2fc_expr)[bad_timing])
colnames(log2fc_expr)[bad_timing]<- gsub("_J8","_J6",  colnames(log2fc_expr)[bad_timing])
colnames(log2fc_expr)[bad_timing]<- gsub("_J7","_J6",  colnames(log2fc_expr)[bad_timing])

j6_samples <- which(! colnames(log2fc_expr) %in% edesign$Description )


log2fc_expr <- log2fc_expr[,which(colnames(log2fc_expr) %in% edesign$Description)]

```

so, data left with 31 Samples, of ctrl and J0-J1-J3
Must be true : 

```{r}

all(colnames(log2fc_expr) %in% edesign$Description)
all(edesign$Description  %in% colnames(log2fc_expr) )
 
 
##> Order Expr-columns as edesign Lines : 
edesign_sample_order <- edesign$Description[which(edesign$Description %in% colnames(log2fc_expr))]

log2fc_expr <- log2fc_expr[,edesign_sample_order]
  
rownames(as.data.frame(edesign) ) <- edesign$Description 

#!  Gives error: rownames Deprecated in tibble, but rownames are assigned whatsoever 

lps2ns_data <- list("edesign"=edesign ,"expr"=log2fc_expr )

```

Function for subsetting data to work only on one pair of comparisons, declare Mort=1, or Viv=1 
 BUT: no need after using transformed data of LPS/NS -> we only have Mort and Viv comparison.
 
```{r}

# subset_conditions <- function(edesign, expr.dat, group, cols2rmv) {
#   
#   #> re-write function with dplyr verbs !
#   group.design <- edesign %>% select(contains(group))
# 
#   sub.design <- edesign %>%
#                 filter(group.design == 1 ) %>%  ## keep only Mort ==1, and
#                 select(-contains(cols2rmv))     ##  remove Viv, because not needed, all ==0
# 
#   ## keep expr that corresponds to same samples as in sub.design :
#   sub.expr <- expr.dat %>% select(matches( rownames(sub.design)))
# 
#   return(list("edesign"=sub.design, "expr"=sub.expr ))
# }

```


```{r}
# maSigPro(data, edesign, matrix = "AUTO", groups.vector = NULL, 
#          degree = 2, time.col = 1, repl.col = 2, group.cols = c(3:ncol(edesign)), 
#          Q = 0.05, alfa = Q, nvar.correction = FALSE, step.method = "backward", rsq = 0.7,
#          min.obs = 3, vars = "groups", significant.intercept = "dummy", cluster.data = 1, 
#          add.IDs = FALSE, IDs = NULL, matchID.col = 1, only.names = FALSE, k = 9,  
#          cluster.method = "hclust", distance = "cor", agglo.method = "ward.D", iter.max = 500, 
#          summary.mode = "median", color.mode = "rainbow", trat.repl.spots = "none",
#          index = IDs[, (matchID.col + 1)], match = IDs[, matchID.col], rs = 0.7, 
#          show.fit = TRUE, show.lines = TRUE, pdf = TRUE, cexlab = 0.8, 
#          legend = TRUE, main = NULL, ...)


## cad : 
# edesign 	: matrix of experimental design. Row names must contain arrayIDs 
# matrix 	 : design matrix for regression analysis


## core functions:
# make.design.matrix(), p.vector(), T.fit(),
# get.siggenes()  and  see.genes()


```

Repeat following for each List

```{r,masigpro_model}


library(maSigPro); library(dplyr); library(ggpubr)

design.msp <- as.data.frame(lps2ns_data$edesign)
expr    <- as.data.frame(lps2ns_data$expr)

## rownames(expr) <- affy_expr$EntrezGeneID # no need of rownames in expr.

## MUST BE TRUE: all(colnames(lps2ns_data$expr) %in% rownames(edesign) )
## Remove 1st col of edesign: "Description", as it is assigned in rownames 
mydesign.mat <- make.design.matrix(design.msp[,-1] ,degree = 3 )

##> see which comparisons will be done
print(mydesign.mat$groups.vector)                         

rownames(mydesign.mat$dis) <- colnames(expr)

## Put expr.data in order of edesign :
#expr <- expr[, rownames(mydesign.mat$dis)]

fit <- p.vector(expr, mydesign.mat, Q=0.05, MT.adjust ="BH", min.obs = 10, counts=TRUE)
# p.vector()returns a list of values:
# > fit$i # returns the number of significant genes
# > fit$alfa # gives p-value at the Q false discovery control level
# > fit$SELEC # is a matrix with the significant genes and their expression values

nb_sigGenes <- fit$i

cat("\nSignificant Genes found: ", nb_sigGenes)
##> 3.3/ Finding significant differences
tstep <- T.fit(fit, step.method = "backward", alfa = 0.05)
                # step.method = "forward"

#! Warning message:
#   In min(nchar(groups.vector)) : no non-missing arguments to min; returning Inf
names(tstep)
### For each selected gene the following values are given:
# >> p-value of the regression ANOVA
# >> R-squared of the model
# >> p-value of the regression coefficients of the selected variables
sigs <- get.siggenes(tstep, rsq = 0.6, vars = "groups")
#rsq = 0.6 :is a cut-off value for the R-squared of the regression model

names(sigs$sig.genes)

##> 4.2/ PLOT:: Visualize  the significant genes in the previous step in VivantvsMort
source("see.genes.mk_function.R")
#source("see.genes.git.R")
sig_data <- sigs$sig.genes
comparison <- names(sigs$sig.genes)[2]
clust_num=9
clustering_genes <- see.genes.mk( sig_data[[comparison]], 
                                  group.cols = c(3:ncol(edesign)), 
                                  cluster.method="hclust", k = clust_num,
                                  dis = mydesign.mat$dis,
                                  show.fit = F,summary.mode = "median",
                                  cluster.data = 1 ) 
                                  
##> Plot All Clusters in a Grid
library(gridExtra)

do.call("grid.arrange", clustering_genes$plots_list  ) # list(..,ncol = clust_num/2) 


```

3.2/ Find significant genes and patterns per time-point for each comparison

Explore output ...

```{r,convert_ids_funct}
##> Find for each Gene their assigned Cluster Number

convert_Entrez2symbol <- function(id_list){
  
      library(org.Hs.eg.db) # installed via biocLite. Cannot call library from an R-variable/object
      require(AnnotationDbi)
      gene_symbols <- mapIds(x=org.Hs.eg.db,
                             keys=id_list,
                             column="SYMBOL",
                             keytype="ENTREZID",
                             multiVals = "first")
      return(gene_symbols)
}

```

Check gene Names in each cluster ...

```{r,find_genes}

## Extract geneSymbols from the IDs that are found for each cluster :
## clustering_genes$cut : gives the genes with their cluster_number (otherwise pluck (clustering_genes,"cut"))
## To find for a specific cluster the geneIDs :
clustering_genes$cut == "1" #  =="2",etc.
## which : gives only the positions where this condition is TRUE
which(clustering_genes$cut == "1")
## names() : gives names of a vector (what is shown above the values as label )
which(clustering_genes$cut == "1") %>%  names()
## Run the commands consecutively to obtain geneNames using "%>%" (pipe)

geneNames_per_cluster <-  function(clust_g,i) { 
          #clust_g %>% glimpse
          which(clust_g$cut  == as.character(i) ) %>%
          names %>% 
          sub("X*","",.) %>% 
          convert_Entrez2symbol 
}
##>output names in EntrezID

## Using function map() we ll run all numbers of clusters 1:9 on the function geneNames_per_cluster (.x takes value from 1:9)
clust_num= length(clustering_genes$plots_list)
geneNames_all_clusters  <- purrr::map(seq(from=1,to=clust_num) , ~geneNames_per_cluster(clustering_genes,.x))

##> Outputs are lists with each element the cluster 1:9, and content the geneNames
geneNames_all_clusters[[1]]

library(glue)
names(geneNames_all_clusters)  <- glue("cluster_{as.character(seq(1,clust_num))}")


## Create Tables with 2 columns : 1/cluster-number and 2/ genes --> save in file

table_geneNames_per_cluster <- map_df(as.character( seq(1,clust_num )), 
                                     ~tibble("cluster"=.x,"genes"=
                                     paste(geneNames_all_clusters[[as.integer(.x)]],collapse=",") ))

# overview:
table_geneNames_per_cluster

```
